File: src/storage/Storage.js
const FileSystem = require('../models/FileSystem');
const File = require('../models/File');
const Folder = require('../models/Folder');

/**
 * Storage class manages the in-memory file system data
 * Migrated from C++ Storage.h/Storage.cpp
 */
class Storage {
  constructor() {
    this.tree = new Map();
    this.fileSystem = new FileSystem();
    this.folders = new Map();
    this.files = new Map();
    
    // Initialize root folder
    this.fileSystem.addFolderId('F0');
    this.folders.set('F0', null);
    this.tree.set('F0', new Map());
    
    // Create base folder
    const baseFolder = new Folder('F1', 'BaseFolder', 'F0');
    this.fileSystem.addFolderId('F1');
    this.folders.set(baseFolder.getId(), baseFolder);
    this.tree.get('F0').set('F1', 1);
  }
  
  /**
   * Add content to a file
   * @param {string} fileName - Name of the file
   * @param {string} content - Content to add
   */
  addContent(fileName, content) {
    const currentFolderId = this.getCurrentFolderId();
    if (this.validateFile(fileName)) {
      const treeFolder = this.tree.get(currentFolderId);
      for (const [key, value] of treeFolder.entries()) {
        if (key[0] === 'f' && this.files.get(key).getFileName() === fileName) {
          this.files.get(key).setContent(content);
        }
      }
    }
  }
  
  /**
   * Generate a new file ID
   * @returns {string} - New file ID
   */
  getNewFileId() {
    return `f${this.files.size}`;
  }
  
  /**
   * Add a file to the file system
   * @param {string} name - File name
   * @param {string} folderId - Parent folder ID
   */
  addFile(name, folderId) {
    const treeFolder = this.tree.get(folderId);
    if (!treeFolder) return;
    
    // Check if file name already exists in folder
    for (const [key, value] of treeFolder.entries()) {
      if (key[0] === 'f' && this.files.get(key).getFileName() === name) {
        console.log("File name already exists! Change the name of the file.");
        return;
      }
    }
    
    // Create new file
    const newFileId = this.getNewFileId();
    const file = new File(newFileId, name, folderId);
    this.files.set(newFileId, file);
    treeFolder.set(newFileId, 1);
    
    return file;
  }
  
  /**
   * Generate a new folder ID
   * @returns {string} - New folder ID
   */
  getNewFolderId() {
    return `F${this.folders.size}`;
  }
  
  /**
   * Add a folder to the file system
   * @param {string} name - Folder name
   * @param {string} parentFolderId - Parent folder ID
   */
  addFolder(name, parentFolderId) {
    const treeFolder = this.tree.get(parentFolderId);
    if (!treeFolder) return;
    
    // Check if folder name already exists in parent folder
    for (const [key, value] of treeFolder.entries()) {
      if (key[0] === 'F' && this.folders.get(key).getName() === name) {
        console.log("Folder name already exists! Change the name of the folder.");
        return;
      }
    }
    
    // Create new folder
    const newFolderId = this.getNewFolderId();
    const folder = new Folder(newFolderId, name, parentFolderId);
    this.folders.set(newFolderId, folder);
    treeFolder.set(newFolderId, 1);
    this.tree.set(newFolderId, new Map());
    
    return folder;
  }
  
  /**
   * Get a folder by ID
   * @param {string} id - Folder ID
   * @returns {Folder} - Folder object
   */
  getFolder(id) {
    return this.folders.get(id);
  }
  
  /**
   * Get a file by ID
   * @param {string} id - File ID
   * @returns {File} - File object
   */
  getFile(id) {
    return this.files.get(id);
  }
  
  /**
   * Get the full path for a folder
   * @param {string} id - Folder ID
   * @returns {string} - Full path
   */
  getPath(id) {
    let folder = this.folders.get(id);
    let path = '';
    
    while (folder && folder.getParentId() !== 'F0') {
      path = '/' + path;
      path = folder.getName() + path;
      folder = this.folders.get(folder.getParentId());
    }
    
    return path || '/';
  }
  
  /**
   * Show folder path
   * @param {string} id - Folder ID
   * @returns {string} - Full path
   */
  showFolderPath(id) {
    return this.getPath(id);
  }
  
  /**
   * Show file path
   * @param {string} id - File ID
   * @returns {string} - Full path
   */
  showFilePath(id) {
    return this.getPath(id);
  }
  
  /**
   * Get the current folder ID
   * @returns {string} - Current folder ID
   */
  getCurrentFolderId() {
    return this.fileSystem.getCurrentFolder();
  }
  
  /**
   * Show all items in a folder
   * @param {string} folderId - Folder ID
   * @returns {Array} - Array of items in the folder
   */
  showItemsInFolder(folderId) {
    const items = [];
    
    if (this.folders.has(folderId)) {
      const treeFolder = this.tree.get(folderId);
      if (treeFolder) {
        for (const [key, value] of treeFolder.entries()) {
          if (key[0] === 'f') {
            items.push({
              type: 'file',
              name: this.files.get(key).getFileName(),
              id: key
            });
          } else {
            items.push({
              type: 'folder',
              name: this.folders.get(key).getName(),
              id: key
            });
          }
        }
      }
    }
    
    return items;
  }
  
  /**
   * Navigate into a folder
   * @param {string} name - Folder name
   * @returns {boolean} - True if successful, false otherwise
   */
  getIntoFolder(name) {
    const currentFolderId = this.fileSystem.getCurrentFolder();
    
    if (name !== '..') {
      const treeFolder = this.tree.get(currentFolderId);
      if (!treeFolder) return false;
      
      for (const [key, value] of treeFolder.entries()) {
        if (key[0] === 'F' && this.folders.get(key).getName() === name) {
          this.fileSystem.addFolderId(key);
          return true;
        }
      }
    } else if (!this.fileSystem.checkEmpty()) {
      this.fileSystem.removeCurrentFolder();
      return true;
    }
    
    return false;
  }
  
  /**
   * Validate if a folder exists in the current directory
   * @param {string} folderName - Folder name
   * @returns {boolean} - True if folder exists, false otherwise
   */
  validateFolder(folderName) {
    const currentFolderId = this.fileSystem.getCurrentFolder();
    const treeFolder = this.tree.get(currentFolderId);
    if (!treeFolder) return false;
    
    for (const [key, value] of treeFolder.entries()) {
      if (key[0] === 'F' && this.folders.get(key).getName() === folderName) {
        return true;
      }
    }
    
    return false;
  }
  
  /**
   * Remove a file from the file system
   * @param {string} fileName - File name
   * @returns {boolean} - True if successful, false otherwise
   */
  removeFile(fileName) {
    if (this.validateFile(fileName)) {
      const currentFolderId = this.fileSystem.getCurrentFolder();
      const treeFolder = this.tree.get(currentFolderId);
      if (!treeFolder) return false;
      
      for (const [key, value] of treeFolder.entries()) {
        if (key[0] === 'f' && this.files.get(key).getFileName() === fileName) {
          const fileId = this.files.get(key).getId();
          this.files.delete(fileId);
          treeFolder.delete(fileId);
          
          if (treeFolder.size === 0) {
            this.tree.delete(currentFolderId);
          }
          
          return true;
        }
      }
    }
    
    return false;
  }
  
  /**
   * Recursively remove a folder and its contents
   * @param {string} node - Folder ID
   */
  removeDFS(node) {
    const treeFolder = this.tree.get(node);
    if (!treeFolder) return;
    
    const itemsToDelete = Array.from(treeFolder.keys());
    
    for (const key of itemsToDelete) {
      if (key[0] === 'F') {
        this.removeDFS(key);
      } else {
        this.files.delete(key);
      }
    }
    
    this.folders.delete(node);
    this.tree.delete(node);
  }
  
  /**
   * Remove a folder from the file system
   * @param {string} folderName - Folder name
   * @returns {boolean} - True if successful, false otherwise
   */
  removeFolder(folderName) {
    if (this.validateFolder(folderName)) {
      const currentFolderId = this.fileSystem.getCurrentFolder();
      const treeFolder = this.tree.get(currentFolderId);
      if (!treeFolder) return false;
      
      for (const [key, value] of treeFolder.entries()) {
        if (key[0] === 'F' && this.folders.get(key).getName() === folderName) {
          const folderId = this.folders.get(key).getId();
          const parFolderId = this.folders.get(key).getParentId();
          
          const parentFolder = this.tree.get(parFolderId);
          if (parentFolder) {
            parentFolder.delete(folderId);
          }
          
          this.removeDFS(folderId);
          return true;
        }
      }
    }
    
    return false;
  }
  
  /**
   * Recursively build folder tree structure
   * @param {string} node - Starting node
   * @param {Object} result - Result object to build
   * @returns {Object} - Tree structure
   */
  buildDFS(node) {
    let result;
    
    if (node[0] === 'F') {
      const folder = this.folders.get(node);
      if (!folder) return null;
      
      result = {
        id: node,
        name: folder.getName(),
        type: 'folder',
        children: []
      };
      
      const treeFolder = this.tree.get(node);
      if (treeFolder) {
        for (const [key, value] of treeFolder.entries()) {
          const childNode = this.buildDFS(key);
          if (childNode) {
            result.children.push(childNode);
          }
        }
      }
    } else {
      const file = this.files.get(node);
      if (!file) return null;
      
      result = {
        id: node,
        name: file.getFileName(),
        type: 'file'
      };
    }
    
    return result;
  }
  
  /**
   * Get folder tree structure
   * @returns {Object} - Tree structure
   */
  getFolderTree() {
    const currentFolderId = this.fileSystem.getCurrentFolder();
    return this.buildDFS(currentFolderId);
  }
  
  /**
   * Validate if a file exists in the current directory
   * @param {string} fileName - File name
   * @returns {boolean} - True if file exists, false otherwise
   */
  validateFile(fileName) {
    const currentFolderId = this.fileSystem.getCurrentFolder();
    const treeFolder = this.tree.get(currentFolderId);
    if (!treeFolder) return false;
    
    for (const [key, value] of treeFolder.entries()) {
      if (key[0] === 'f' && this.files.get(key).getFileName() === fileName) {
        return true;
      }
    }
    
    return false;
  }
}

// Singleton pattern
let instance = null;

module.exports = {
  getInstance: () => {
    if (instance === null) {
      instance = new Storage();
    }
    return instance;
  },
  resetInstance: () => {
    instance = null;
    return module.exports.getInstance();
  }
};

File: src/storage/Storage.cpp
// src/storage/storage.cpp

#include "../../include/storage/Storage.h"
#include "../../include/models/File.h"
#include "../../include/models/FileSystem.h"
#include "../../include/models/Folder.h"

#include <vector>
#include <string>
#include <map>
#include <iostream>
#include <stack>
#include <queue>
using namespace std;

Storage *Storage::instance = nullptr;

Storage *Storage::getInstance()
{
    if (instance == nullptr)
    {
        return instance = new Storage();
    }
    return instance;
}

Storage::Storage()
{
    fileSystem = new FileSystem();
    fileSystem->addFolderId("F0");
    folders["F0"] = nullptr;
    map<string, int> temp;
    tree["F0"] = temp;
    Folder *f = new Folder("F" + to_string(folders.size()), "BaseFolder", "FX");
    fileSystem->addFolderId("F0");
    fileSystem->addFolderId("F1");
    folders[f->getId()] = f;
}

void Storage::addContent(string fileName, string content)
{
    string currentFolderId = getCurrentFolderId();
    if (validateFile(fileName))
    {
        for (auto i : tree[currentFolderId])
        {
            if (i.first[0] == 'f' && files[i.first]->getFileName() == fileName)
            {
                files[i.first]->setContent(content);
            }
        }
    }
}

string Storage::getNewFileId() { return "f" + to_string(files.size()); }

void Storage::addFile(string name, string folderId)
{
    for (auto i : tree[folderId])
    {
        if (i.first[0] == 'f' && files[i.first]->getFileName() == name)
        {
            cout << "     " << "File name already exist! change the name of the file." << endl;
            return;
        }
    }
    string newFileId = getNewFileId();
    File *f = new File(newFileId, name, folderId);
    files[newFileId] = f;
    tree[folderId][newFileId] = 1;
    cout << "     " << "File created! File name = " + name + ", id =" + f->getId() + ", in folder id - " << folderId << endl;
}

string Storage::getNewFolderId() { return "F" + to_string(folders.size()); }

void Storage::addFolder(string name, string parentFolderId)
{
    for (auto i : tree[parentFolderId])
    {
        if (i.first[0] == 'F' && folders[i.first]->getName() == name)
        {
            cout << "     " << "Folder name already exist! change the name of the folder." << endl;
            return;
        }
    }
    string newFolderId = getNewFolderId();
    Folder *f = new Folder(newFolderId, name, parentFolderId);
    folders[newFolderId] = f;
    tree[parentFolderId][newFolderId] = 1;
    cout << "     " << "New folder created! Name = " << name << " id = " << f->getId() << endl;
}

Folder *Storage::getFolder(string id)
{
    return folders[id];
}

File *Storage::getFile(string id)
{
    return files[id];
}

string Storage::getPath(string id)
{
    Folder *f = folders[id];
    string path = "";
    while (f && f->getParentId() != "F0")
    {
        path = "/" + path;
        path = f->getName() + path;
        f = folders[f->getParentId()];
    }
    return path;
}

void Storage::showFolderPath(string id)
{
    string path = getPath(id);
    cout << path << endl;
}

void Storage::showFilePath(string id)
{
    string path = getPath(id);
    cout << path << endl;
}

string Storage::getCurrentFolderId() { return fileSystem->getCurrentFolder(); }

void Storage::showItemsInFolder(string folderId)
{
    if (folders.count(folderId) && folders[folderId])
    {
        for (auto i : tree[folderId])
        {
            if (i.first[0] == 'f')
                cout << "     " << files[i.first]->getFileName() << endl;
            else
                cout << "     " << folders[i.first]->getName() << endl;
        }
    }
    else
        cout << "     " << "Folder does not exist." << endl;
}

void Storage::getIntoFolder(string name)
{
    string currentFolderId = fileSystem->getCurrentFolder();
    if (name != "..")
    {
        for (auto i : tree[currentFolderId])
        {
            if (i.first[0] == 'F' && folders[i.first]->getName() == name)
            {
                fileSystem->addFolderId(i.first);
                return;
            }
        }
    }
    else if (!fileSystem->checkEmpty())
    {
        fileSystem->removeCurrentFolder();
        return;
    }
    cout << "     " << "Wrong file name, no file exists with name " << name << endl;
}

bool Storage::validateFolder(string folderName)
{
    string currentFolderId = fileSystem->getCurrentFolder();
    for (auto i : tree[currentFolderId])
    {
        if (i.first[0] == 'F' && folders[i.first]->getName() == folderName)
        {
            return true;
        }
    }
    return false;
}

void Storage::removeFile(string fileName)
{
    if (validateFile(fileName))
    {
        string currentFolderId = fileSystem->getCurrentFolder();
        for (auto i : tree[currentFolderId])
        {
            if (i.first[0] == 'f' && files[i.first]->getFileName() == fileName)
            {
                string fileId = files[i.first]->getId();
                files.erase(fileId);
                tree[currentFolderId].erase(fileId);
                if (tree[currentFolderId].size() == 0)
                    tree.erase(currentFolderId);
                cout << "File removed successfully!" << endl;
                return;
            }
        }
    }
}

void Storage::removeDFS(string node)
{
    for (auto i : tree[node])
    {
        if (i.first[0] == 'F')
        {
            removeDFS(i.first);
        }
        else
        {
            cout << "     " << "File id - " << files[i.first]->getId() << " and name - " << files[i.first]->getFileName() << " removed successfully!" << endl;
            files[i.first] = nullptr;
        }
    }
    cout << "     " << "Folder id - " << folders[node]->getId() << " and name - " << folders[node]->getName() << " removed successfully!" << endl;
    folders[node] = nullptr;
    tree.erase(node);
}

void Storage::removeFolder(string folderName)
{
    if (validateFolder(folderName))
    {
        string currentFolderId = fileSystem->getCurrentFolder();
        for (auto i : tree[currentFolderId])
        {
            if (i.first[0] == 'F' && folders[i.first]->getName() == folderName)
            {
                string folderId = folders[i.first]->getId();
                string parFolderId = folders[i.first]->getParentId();
                tree[parFolderId].erase(folderId);
                removeDFS(folderId);
                cout << "     Folder removed successfully!" << endl;
                return;
            }
        }
    }
}

void Storage::showDFS(string node, string symbols)
{
    cout << "     " << symbols + "- " << ((node[0] == 'F') ? folders[node]->getName() : files[node]->getFileName()) << endl;

    symbols += "  |";
    for (auto i : tree[node])
    {
        showDFS(i.first, symbols);
    }
}

void Storage::showFolderTree()
{
    string currentFolderId = fileSystem->getCurrentFolder();
    string symbols = "";
    showDFS(currentFolderId, symbols);
}

bool Storage::validateFile(string fileName)
{
    string currentFolderId = fileSystem->getCurrentFolder();
    for (auto i : tree[currentFolderId])
    {
        if (i.first[0] == 'f' && files[i.first]->getFileName() == fileName)
        {
            return true;
        }
    }
    return false;
}

// Grep support methods
vector<string> Storage::getFileIdsInFolder(string folderId)
{
    vector<string> fileIds;
    if (tree.find(folderId) != tree.end())
    {
        for (auto i : tree[folderId])
        {
            if (i.first[0] == 'f')
            {
                fileIds.push_back(i.first);
            }
        }
    }
    return fileIds;
}

vector<string> Storage::getFolderIdsInFolder(string folderId)
{
    vector<string> folderIds;
    if (tree.find(folderId) != tree.end())
    {
        for (auto i : tree[folderId])
        {
            if (i.first[0] == 'F')
            {
                folderIds.push_back(i.first);
            }
        }
    }
    return folderIds;
}

string Storage::getFileIdByName(string fileName, string folderId)
{
    if (tree.find(folderId) != tree.end())
    {
        for (auto i : tree[folderId])
        {
            if (i.first[0] == 'f' && files[i.first] && files[i.first]->getFileName() == fileName)
            {
                return i.first;
            }
        }
    }
    return "";
}

map<string, File*> Storage::getAllFiles()
{
    return files;
}

map<string, Folder*> Storage::getAllFolders()
{
    return folders;
}
